nordvpn: init at 5.2.0

different-error opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Add the popular NordVPN to NixOS. Tested using the following configuration:

```nix
{
  config,
  lib,
  pkgs,
  ...
}:

{
  imports = [
    ./hardware-configuration.nix
  ];

  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;
  networking.firewall.enable = false;  # required

  services.nordvpn.enable = true;  # required

  virtualisation.vmVariant = {
    virtualisation = {
      memorySize = 4096;
      cores = 3;
    };
  };

  users.groups.alice = {};
  users.users.alice = {
    isSystemUser = true;
    password = "alice";
    group = "alice";
    extraGroups = [
      "wheel"
      "nordvpn"  # strongly recommended
    ];
    shell = pkgs.bash;
    home = "/home/alice";
    createHome = true;
    packages = with pkgs; [
      tree
    ];
  };

  system.stateVersion = "24.11"; # Did you read the comment?
}
```

The configuration was tested by running:

```bash
nixos-rebuild build-vm --use-remote-sudo -I nixos-config=/path/to/above/configuration.nix -I nixpkgs=/path/to/this/pr/nixpkgs
```

There is [another PR (#220616) for NordVPN](https://github.com/NixOS/nixpkgs/pull/220616), which is over two years old and has been stale for a year. Additionally, there are issues requesting NordVPN support for NixOS [here](https://github.com/NixOS/nixpkgs/issues/101864) and [here](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945).

I chose to extract the **.deb** package instead of building from source ~~to avoid [modifying or leaking the salt](https://github.com/NordSecurity/nordvpn-linux/blob/main/BUILD.md#building-nordvpn-cli-utility)~~. [Meshnet is not yet supported](https://github.com/NordSecurity/nordvpn-linux/issues/869), but core NordVPN features work. I’ll create another PR once the Meshnet issues are resolved.

![2025-05-13-123306_hyprshot](https://github.com/user-attachments/assets/53a00542-feaa-45af-bc65-3093e5fa23fe)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->


---

Note: I use Claude btw.

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

different-error committed:
maintainers: add different-error

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,83 @@
+
+    enable = lib.mkEnableOption "Enable NordVPN";
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = null;
+      description = ''
+        Use this option if you want to use a custom nordvpn package.
+        Example custom nordvpn package for using an already downloaded .deb:
+        https://gist.github.com/SanferDsouza/956b1842cb4636b649b9937c7f7e93ea
+      '';
+      example = ''
+        pkgs.callPackage /home/user/mycustompkgs/custom-nordvpn.nix
+      '';
+    };
```
```suggestion
      package = lib.mkPackageOption pkgs "nordvpn" { };
```

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,83 @@
+      nordvpnConf = "${cfg.package}/lib/tmpfiles.d/nordvpn.conf";
+    in
+    lib.mkIf cfg.enable {
+      services.nordvpn.package = lib.mkDefault pkgs.nordvpn;
+
```
remove this

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,83 @@
+            Environment = "PATH=/run/current-system/sw/bin";
+            ExecStart = "${nordvpnPkg}/sbin/nordvpnd";
+            ExecStartPre = ''
+              ${pkgs.bash}/bin/bash -c '\
+                if [ -z "$(ls -A /var/lib/nordvpn)" ]; then \
+                  cp -r ${nordvpnPkg}/var/lib/nordvpn/* /var/lib/nordvpn; \
+                fi'
+            '';
```
I don't think this is required.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+{
+  pkgs,
+  system ? stdenv.hostPlatform.system,
```
This should be removed.

SanferDsouza replied:
system is the key to the `archMap`. Facilitates multiple architecture support.

drupol replied:
Yes, that I have seen, thank you :)

There's no need to have it there, move it inside the derivation, don't expose `system` as parameter.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+{
```
Why 2 derivations in here? I think everything can be done within a single one.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
```
If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+    ln -s ${thisPkg}/var       $out/
+    cp -r ${thisPkg}/usr/lib   $out/
+
+    servicePath=$out/lib/systemd/system/nordvpnd.service
+
+    # can have at most one ExecStart, use Exec instead which can be overriden
+    substituteInPlace $servicePath --replace-fail ExecStart Exec
```
Why this in a package?

SanferDsouza replied:
You're right. Instead of doing it this way, I should simply replace the path from `/usr/sbin/nordvpn` to its nix path.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+    substituteInPlace $servicePath --replace-fail ExecStart Exec
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

SanferDsouza replied:
I've seen both in nixpkgs but ok

drupol replied:
True, we try to remove them as much as we can. While this is technically valid, it exposes the whole `lib` in the attribute set, possibly making confusion with LSPs and IDEs.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+  meta = with lib; {
+    description = "CLI client and daemon for NordVPN";
+    homepage = "https://www.nordvpn.com";
+    license = licenses.gpl3;
```
```suggestion
    license = lib.licenses.gpl3;
```

drupol reviewed (changes requested):
Hello,

Thanks for your first PR.

I made some feedback, let me know if you need some help.

NyCodeGHG commented:
For such a security relevant package such as a vpn software, we should build from source if possible.
I'm not sure why you want to avoid "leaking" the salt, since it can be easily extracted from the binary in the .deb you linked.

NyCodeGHG commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+    inherit pname version;
+    src = fetchurl {
+      url = "${baseURL}/${pname}_${version}_${systemArch}.deb";
+      hash = "sha256-2eLVr0aQpMhIMCrkyIRcCabXc197GJZ31Vl5BND7Di8=";
```
this hash will not match for every architecture

NyCodeGHG commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,112 @@
+    ];
+    buildInputs =
+      [ libxml2 ]
+      ++ (with pkgs; [
```
dependencies should be passed via the function definition at the top of the file, not with `pkgs.`

SanferDsouza commented:
> For such a security relevant package such as a vpn software, we should build from source if possible. I'm not sure why you want to avoid "leaking" the salt, since it can be easily extracted from the binary in the .deb you linked.
> 
> This would also allow us to patch out quirks such as relying on /usr

Your claims about salts are valid and correct! My earlier statement was based on an incorrect misunderstanding, and I've updated the comment accordingly.

Another way to address quirks is to make changes directly in the NordVPN repository, which was my implicit assumption for the next release.

I could update this to build from scratch, but I don’t see the added security benefit. All other Linux distributions download from `repo.nordvpn.com` ([.deb](https://repo.nordvpn.com/deb/nordvpn/debian/) or [.rpm](https://repo.nordvpn.com/yum/nordvpn/centos/)) via their [install script](https://downloads.nordcdn.com/apps/linux/install.sh). I’ve also noticed that [ExpressVPN extracts their .deb](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ex/expressvpn/package.nix). Fortunately, building from source is nearly complete thanks to [community efforts](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945). Regardless, I’m not convinced there’s a significant advantage. Lmk your thoughts, thanks!

drupol commented:
> I could update this to build from scratch, but I don’t see the added security benefit. All other Linux distributions download from `repo.nordvpn.com` ([.deb](https://repo.nordvpn.com/deb/nordvpn/debian/) or [.rpm](https://repo.nordvpn.com/yum/nordvpn/centos/)) via their [install script](https://downloads.nordcdn.com/apps/linux/install.sh). I’ve also noticed that [ExpressVPN extracts their .deb](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ex/expressvpn/package.nix). Fortunately, building from source is nearly complete thanks to [community efforts](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945). Regardless, I’m not convinced there’s a significant advantage. Lmk your thoughts, thanks!

You are right that most distributions fetch the `.deb` or `.rpm` packages from `repo.nordvpn.com` using the official install script, and that ExpressVPN follows a similar approach in Nixpkgs by extracting the `.deb` archive.

However, the key issue here is **trust and verifiability**.

At the moment, there's no reliable way to verify that the binaries provided on `repo.nordvpn.com` are actually built from the publicly available sources. By using those prebuilt packages, **we are implicitly trusting the vendor** without any way to independently validate the build integrity.

One of the strengths of Nix is its focus on reproducibility. Building from source allows us (most of the time) to produce reproducible outputs. This enables a verifiable 1-to-1 mapping between the source code and the resulting binaries, which significantly improves the security of the software supply chain.

Fortunately, thanks to recent community efforts, we’re getting close to being able to build the client fully from source.

That’s why I believe it’s worth pushing in that direction.

SanferDsouza commented:
> > I could update this to build from scratch, but I don’t see the added security benefit. All other Linux distributions download from `repo.nordvpn.com` ([.deb](https://repo.nordvpn.com/deb/nordvpn/debian/) or [.rpm](https://repo.nordvpn.com/yum/nordvpn/centos/)) via their [install script](https://downloads.nordcdn.com/apps/linux/install.sh). I’ve also noticed that [ExpressVPN extracts their .deb](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ex/expressvpn/package.nix). Fortunately, building from source is nearly complete thanks to [community efforts](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945). Regardless, I’m not convinced there’s a significant advantage. Lmk your thoughts, thanks!
> 
> You are right that most distributions fetch the `.deb` or `.rpm` packages from `repo.nordvpn.com` using the official install script, and that ExpressVPN follows a similar approach in Nixpkgs by extracting the `.deb` archive.
> 
> However, the key issue here is **trust and verifiability**.
> 
> At the moment, there's no reliable way to verify that the binaries provided on `repo.nordvpn.com` are actually built from the publicly available sources. By using those prebuilt packages, **we are implicitly trusting the vendor** without any way to independently validate the build integrity.
> 
> One of the strengths of Nix is its focus on reproducibility. Building from source allows us (most of the time) to produce reproducible outputs. This enables a verifiable 1-to-1 mapping between the source code and the resulting binaries, which significantly improves the security of the software supply chain.
> 
> Fortunately, thanks to recent community efforts, we’re getting close to being able to build the client fully from source.
> 
> That’s why I believe it’s worth pushing in that direction.

Gotcha. A malicious attacker might *somehow* tamper with their binaries. Building from source is the secure way to go. Ok, will do, thanks!

SanferDsouza commented:
Modified the package to build from source instead of extracting the .deb file.
Attribution: I adapted the working configuration found [here](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945).

Verified that core features function correctly.
![2025-05-13-232059_hyprshot](https://github.com/user-attachments/assets/00f2cfe2-ecf7-4d47-93d9-9041124fe605)

Thank you all for your time!

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,66 @@
+      description = "NordVPN daemon.";
+      name = "nordvpnd.service";
+      serviceConfig = {
+        Environment = "PATH=/usr/lib/nordvpn:/run/current-system/sw/bin";
```
This sounds a bit weird. Is it really needed ?

SanferDsouza replied:
`nordvpnd` calls the `wg` and `openvpn` binaries, located in `/usr/lib/nordvpn`. It also requires several other binaries available in the current system. If preferred, I can add them individually to the `systemd.services.nordvpnd.path` instead (and in the process get rid of relying on the binaries found in `/usr/lib/nordvpn` and `/run/current-system/sw/bin`)?

drupol replied:
I think those binaries should be installed in nordvpn's `$out/bin` directory. If that is not an option, you should then just use `PATH=${nordvpn}/lib/nordvpn` or something like that. Using absolute paths is not a common practice in this project.

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,66 @@
+      "D /run/nordvpn 0770 root nordvpn"
+    ];
+
+    users.groups.nordvpn = { };
```
Is it needed ? Can't we use DynamicUser ?

drupol replied:
Have you tried removing that and using systemd's `DynamicUser` directive? Have a look at other NixOS services for some inspiration.

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,66 @@
+
+  config = lib.mkIf cfg.enable {
+    systemd.tmpfiles.rules = [
+      # nordvpnd relies on hardcorded path to the following
```
Since we are building it from the source, I guess we can change that ?

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,66 @@
+  options.services.nordvpn = {
+
+    enable = lib.mkEnableOption "Enable NordVPN";
+
```
Please avoid adding extra blank lines for no reason. If you really want to keep them, please make this pattern consistent in all the files.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,151 @@
+  sysctl,
+  wireguard-tools,
+  zlib,
+  ...
```
This is not needed and must be removed.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,151 @@
+{
+  pkgs,
```
Don't use pkgs, inject each dependency one by one

NyCodeGHG commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,92 @@
+
+    services.resolved.enable = true;
+
+    security.wrappers.ip = {
+      owner = "root";
+      group = nordvpn;
+      source = "${pkgs.iproute2}/bin/ip";
+      capabilities = "cap_net_admin+eip";
+      permissions = "u+rx,g+rx";
+    };
```
Could you explain why this is required?

SanferDsouza replied:
ah this is not required, will remove and force push

SanferDsouza commented:
I've reduced privileges by using a dedicated `nordvpn` user. `DynamicUser=true` behaved inconsistently when I specified the `nordvpn` group, including when I set it only in `SupplementaryGroups=`. Additionally, `nordvpnd` failed to recognize the `norduserd` process, even though both ran as the same user, which is incorrect behavior. As far as I know, this only affects notifications for VPN server connection/disconnection. I verified basic connect/disconnect operations using OpenVPN and NordLynx protocols.

![2025-05-16-002450_hyprshot](https://github.com/user-attachments/assets/5fcfc03b-db4c-4998-a974-0ca1a29bd83e)

NyCodeGHG commented on doc/release-notes/rl-2505.section.md:
```diff
@@ -224,6 +224,8 @@
 
 - `nodePackages.copy-webpack-plugin` has been removed, as it should be installed in projects that use it instead.
 
+- [nordvpn](https://www.nordvpn.com), one of the fastest VPNs in the world.
```
this doesn't really sound like a changelog entry and more like an ad.

NyCodeGHG commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,90 @@
+  nordvpn = "nordvpn";
+  myPkgs =
+    with pkgs;
+    [
+      iptables
+      iproute2
+      procps
+      sysctl
+      e2fsprogs
+    ]
```
why are these installed to systemPackages?

NyCodeGHG commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,90 @@
+    users.users."${nordvpn}" = {
+      isNormalUser = true;
+      group = nordvpn;
+      home = "/home/${nordvpn}";
```
Why is this not a system user?
And is the nordvpn variable used when it's just "nordvpn"?
The home directory probably also shouldn't be at /home

NyCodeGHG commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,90 @@
+        RuntimeDirectory = nordvpn;
+        RuntimeDirectoryMode = "0750";
+        StateDirectory = nordvpn;
+        SupplementaryGroups = [ nordvpn ];
```
This shouldn't be required since it's already running as the nordvpn group

NyCodeGHG commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,90 @@
+    systemd.services.nordvpnd = {
+      after = [ "network-online.target" ];
+      description = "NordVPN daemon.";
+      name = "nordvpnd.service";
```
This is the default value and not required

NyCodeGHG commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,150 @@
+    mkdir -p $BIN_PATH
+    mv bin/norduserd $BIN_PATH
+    ln -s ${patchedPkgs.openvpn}/bin/openvpn $BIN_PATH
+    ln -s ${wireguard-tools}/bin/wg $BIN_PATH
```
This leads to leaking the patched openvpn and wireguard-tools into the bin directory for the end user.
Can we patch nordvpn to give them a full path to the binary directly?

NyCodeGHG commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,150 @@
+
+  nativeBuildInputs = [
+    pkg-config
+    gcc
```
buildGoModule includes stdenv by default, so there's a c compiler available.
This shouldn't be required

SanferDsouza commented:
I've updated the source to use a hard-coded path to the patched `openvpn` and removed it from `bin/`. I've also removed `wg` from `bin/` and instead added `wireguard-tools` to the `PATH` environment of `nordvpnd.service`.

Additionally, I've added a `norduserd` user `systemd` service, which appears to correctly fix nordvpn notifications.

![2025-05-16-165342_hyprshot](https://github.com/user-attachments/assets/acac0c9f-b72a-4234-a7c5-79aaa56d0a69)

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,99 @@
+let
+  cfg = config.services.nordvpn;
+  nordvpnPkg = cfg.package;
+  nordvpn = "nordvpn";
```
Please remove this, use string literal.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,148 @@
+    })
+  ];
+in
+buildGoModule rec {
```
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

And then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent examples of implementations: https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aclosed+finalAttrs+pattern+

Feel free to reach out if you need some assistance.

SanferDsouza replied:
I'll just use multiple `let .. in` statements since it's easier to understand.

drupol replied:
Please use finalattrs pattern. There's no need of let in.

SanferDsouza replied:
I was able to shrink it down to one `let .. in` without `rec`. lmk if that works!

drupol replied:
OK this is your first PR here and you're already not listening to the advice from the people contributing from years. I'll unsubscribe and leave this PR for someone else to review from now on.

drupol replied:
Yes, it's ok . I deleted my previous comment, ignore it.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,148 @@
+  src = fetchFromGitHub {
+    owner = "NordSecurity";
+    repo = "nordvpn-linux";
+    rev = version;
```
```suggestion
    tag = version;
```

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,148 @@
+    owner = "NordSecurity";
+    repo = "nordvpn-linux";
+    rev = version;
+    sha256 = "sha256-XB/wOdp97VnAkzHTHrJ2xffEx4HNFWn9yjFz3TsaEsE=";
```
```suggestion
    hash = "sha256-XB/wOdp97VnAkzHTHrJ2xffEx4HNFWn9yjFz3TsaEsE=";
```

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,148 @@
+  buildPhase = ''
+    runHook preBuild
+
+    echo "Building nordvpn CLI..."
```
I understand that these `echo` might be useful during the making of this derivation, but I believe they should be all removed.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,148 @@
+    sed -i "s|$old_ovpn_path|$new_ovpn_path|" daemon/vpn/openvpn/config.go
+
+    export LDFLAGS="${builtins.concatStringsSep " " ldflags}"
+    go build -ldflags "$LDFLAGS" -o bin/nordvpn ./cmd/cli
```
The `buildGoModule` wrapper is made to build go app. I see that here you're overriding the `buildPhase`, hindering the wrapper mechanism. Wouldn't it be possible to avoid using the `buildPhase` and use exclusively `buildGoModule`?

NyCodeGHG commented:
Please don't be offended if my guess is wrong.
Did you use AI/LLMs to write this pull request?

Considering this is your first PR, a lot of stuff just feels very *off*.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,143 @@
+  postInstall = ''
+    # rename to standard names
+    BIN_DIR=$out/bin
+    mv $BIN_DIR/daemon $BIN_DIR/nordvpnd
```
Can you use `install` with correct flags instead ?

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,143 @@
+    # Nordvpn needs icons for the system tray and notifications
+    ASSETS_PATH=$out/share/icons/hicolor/scalable/apps
+    mkdir -p $ASSETS_PATH
+    cp assets/icon.svg $ASSETS_PATH/nordvpn.svg
```
Same here, use install. There's no need to use mkdir when using it, it is lesser code to maintain.

drupol commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,143 @@
+    done
+  '';
+
+  doCheck = false;
```
Please, explain in a comment why this is disabled.

SanferDsouza replied:
I'll update to run a subset of unit tests that do not require network access.

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,98 @@
+      after = [ "network.target" ];
+      wantedBy = [ "default.target" ];
+      serviceConfig = {
+        ExecStart = "${nordvpnPkg}/bin/norduserd";
```
You could use lib.getExe'

drupol replied:
Just like you did in `systemd.services.nordvpnd`

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,98 @@
+        RuntimeDirectory = "nordvpn";
+        RuntimeDirectoryMode = "0750";
+        StateDirectory = "nordvpn";
+        User = "nordvpn";
```
Have you tried with DynamicUser ?

SanferDsouza replied:
Previously, I tried the following configuration:

```
DynamicUser = true;
SupplementaryGroups = [ "nordvpn" ];
```

This required `sudo nordvpn c` because the dynamic user `nordvpnd` was not recognized as part of the `nordvpn` group in the system’s group database, preventing access to the NordVPN socket.

I also attempted:

```
DynamicUser = true;
Group = "nordvpn";
```

This required a static `nordvpn` group, making a static user preferable. Additionally, `DynamicUser` caused DNS resolution failures ("Transport endpoint is not connected"), likely due to Polkit-related policy issues, resulting in service failures.

SanferDsouza commented:
The NordVPN package build fails because [these changes](https://github.com/NixOS/nixpkgs/pull/406725/commits/c46d8df628fc742ca6d50419635a2fc1e2f5f508) have not been merged into master. Updated `checkPhase` to skip network-dependent tests (`TestTransports`, `TestH1Transport_RoundTrip`) due to sandbox restrictions.

Thanks again for the review!

![2025-05-17-122009_hyprshot](https://github.com/user-attachments/assets/c891b146-539f-4ad0-b932-8af345dba95f)

drupol commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -3,7 +3,11 @@
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
-- Create the first release note entry in this section!
+## New Modules {#sec-nixpkgs-release-25.11-new-modules}
```
2505 hasn't been released yet, are you sure this is the proper place to add this?

Perhaps @leona-ya can help?

leona-ya replied:
Depends on when this will be merged. When this will be merged and backported to 25.05 before release, 25.05 is the right release notes file.

Side-note: this is not the right release-notes file: This is the Nixpkgs release-notes file, not the NixOS.

GetPsyched replied:
Also, as per the error in the CI, this section needs to be added to the redirects (command to do so also present in the CI)

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,99 @@
+      serviceConfig = {
+        AmbientCapabilities = "CAP_NET_ADMIN";
+        CapabilityBoundingSet = "CAP_NET_ADMIN";
+        ExecStart = ''${lib.getExe' nordvpnPkg "nordvpnd"}'';
```
```suggestion
        ExecStart = lib.getExe' nordvpnPkg "nordvpnd";
```

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,99 @@
+
+    users.groups.nordvpn = { };
+
+    users.users.nordvpn = {
```
I see that the user and group `nordvpn` are hardcoded. Don't you think we should let the end-user decide which user and group to set? Perhaps you should create two new config options `user` and `group` with their default value set to `nordvpn`.

drupol commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,99 @@
+
+let
+  cfg = config.services.nordvpn;
+  nordvpnPkg = cfg.package;
```
This can be removed. Use `cfg.package`.

drupol reviewed (commented):
I made a few inline remarks.

Also:

1. Avoid adding extra blank lines (e.g., after a `let ... in` statement).
2. In the PR, please mark as resolved the comments that are done

SanferDsouza commented:
I forgot to test the standalone package. It now works as expected. Users of the standalone package would need to run `sudo nordvpnd` somehow and then can connect to the vpn using `nordvpn c`. Using the `services.nordvpn` option is the preferred solution for nixos.

drupol reviewed (commented):
The commit https://github.com/NixOS/nixpkgs/pull/406725/commits/21b0a13e9e2f47c4b39cf9045acd87959e6bd58a (`nordvpn: init at 3.20.2`) contains module changes that should not belong to that commit.

SanferDsouza commented:
Tested and working using openvpn and nordlynx technologies. Meshnet is not working yet. Meshnet is out of scope of this PR.

SanferDsouza commented on nixos/modules/services/networking/nordvpn.md:
````diff
@@ -0,0 +1,56 @@
+```
+
+Currently, the NixOS firewall must be disabled to connect to the NordVPN service,
+as its rules conflict with NordVPN’s firewall requirements.
````
Well, I don't fully understand it just yet either. But the nixos firewall adds some rules to the firewall on its own (i.e. it does it without the user specifying those rules in their configuration). I tried a few things like opening ports and adding forwarding rules, but wasn't able to get it to work.

Others in the community have expressed similar sentiment [here](https://github.com/NixOS/nixpkgs/issues/101864#issuecomment-1514360289) and [here](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2486424885).

I'm planning to give it another shot this weekend! But I'd like this PR merged to get early feedback asap.

Blatzar replied:
By extensively checking my own working config I was able to narrow down the fix to: Setting [networking.firewall.checkReversePath](https://mynixos.com/nixpkgs/option/networking.firewall.checkReversePath) to false. This way I can keep my firewall on while using nordvpn. I do not know the security implications of this, but it is worth further investigating.

ruffsl replied:
I think I've seen and encountered that before when using just wire guard:

https://discourse.nixos.org/t/solved-wireguard-and-firewall-weirdness/58126/4

My understanding is that NordVPN also makes use of wire guard as well, no?

Blatzar commented:
> Please don't be offended if my guess is wrong.
> Did you use AI/LLMs to write this pull request?
> Considering this is your first PR, a lot of stuff just feels very off.

A lot of the weird mistakes probably come from my [initial build](https://github.com/NordSecurity/nordvpn-linux/issues/355#issuecomment-2480381945). It was made to build by any means necessary, which meant using some (probably) very weird first working solutions I could find on google (like buildFHSEnv). buildGoModule had `buildGoModule rec {` in the [example](https://nixos.wiki/wiki/Go) so that was used, and so on.

The author is likely not using AI, it was just me building a very amateurish base. Sorry :sweat:

SanferDsouza commented:
Rebase over latest master and use refactored flags in tests

```
  services.displayManager.gdm.enable = true;
  services.desktopManager.gnome.enable = true;
```

SanferDsouza commented:
Updated documentation. Verified working with `networking.firewall.checkReversePath = "loose";`.

ruffsl commented on nixos/modules/services/networking/nordvpn.nix:
```diff
@@ -0,0 +1,142 @@
+          Group = cfg.group;
+          KillMode = "process";
+          NonBlocking = true;
+          Requires = "nordvpnd.socket";
```
Where is the definition for this `nordvpnd.socket`? Must that systemd unit also be declared here in this same service module file? I.e. is there TCP port or unit socket path is this .socket mapped to?

SanferDsouza replied:
ahh I forget there was the `.socket` definition, thanks for pointing it out, will add soon

```
[Unit]
Description=NordVPN Daemon Socket
PartOf=nordvpnd.service

[Socket]
ListenStream=/run/nordvpn/nordvpnd.sock
NoDelay=true
# SocketUser=root
SocketGroup=nordvpn
SocketMode=0770
DirectoryMode=0750
NoDelay=true

[Install]
WantedBy=sockets.target
```

ruffsl commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,168 @@
+    sed -i "s|$old_ovpn_path|$new_ovpn_path|" daemon/vpn/openvpn/config.go
+
+    # fix missing include in a broken module preventing cgo compilation
+    chmod +w vendor/github.com/jbowtie/gokogiri/xpath/
```
Could you double check this? I encountered permission issues when attempting to `nix build .#nordvpn` this locally. Ended up resorting to `chmod -R +w vendor/github.com/jbowtie/gokogiri/xpath` to get the build to work for the first time before reverting the change.

ruffsl replied:
I think something like this using `modPostBuild` to patch the vendor folder would be more pure:

- d1d2b928be76a0cdfb96ef0cfa22ded1f6eaf94d

SanferDsouza replied:
Thanks for the patch. I'm surprised that `chmod +w` doesn't work but `chmod -R +w` works. I've updated the **package.nix** to apply the patch. Since I cannot reproduce the issue on my end, could you try the latest changes and verify that it works as expected?

ruffsl replied:
I think the vindor folding symlinks to module paths in the nix store, so while chmod works with sudo nix switch, it doesn't work when nix building without root, nor was it pure with respect to the state of the nix store.

ruffsl commented:
Had a little trouble building building this locally at first, but does seem to run fine using the README instructions:

```shell
$ nordvpn login --token ...
Welcome to NordVPN! You can now connect to VPN by using 'nordvpn connect'.

NOTE: By default, all users who are members of the 'nordvpn' group have permission to control the NordVPN application.
To limit access exclusively to the root user, remove all users from the 'nordvpn' group.

$ nordvpn c
Connecting to France #746 (fr746.nordvpn.com)
You are connected to France #746 (fr746.nordvpn.com)!

$ ip address
...
15: nordlynx: <POINTOPOINT,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none 
    inet ... scope global nordlynx
       valid_lft forever preferred_lft forever
```

---

Any chance of registering the nordvpn binary with the desktop `nordvpn://` URI, to simplify browser based oauth?

```
$ nordvpn login
Continue in the browser: https://api.nordvpn.com/v1/users/oauth/login-redirect?attempt=...
```

KDE Popup:
```
Could not read file nordvpn://login?action=login&exchange_token=...
```

Related?

- https://github.com/youssefaltai/linux-custom-url-scheme
- https://askubuntu.com/questions/514125/url-protocol-handlers-in-basic-ubuntu-desktop
- https://github.com/NixOS/nixpkgs/issues/18791

dimkNevidimk commented:
and in case someone else stubmles into it - if you use powerlevel10k with zsh just adding nordvpn will cause it to hang forever:
```
+prompt_nordvpn:1> unset P9K_NORDVPN_STATUS P9K_NORDVPN_TECHNOLOGY P9K_NORDVPN_PROTOCOL P9K_NORDVPN_IP_ADDRESS P9K_NORDVPN_SERVER P9K_NORDVPN_COUNTRY P9K_NORDVPN_CITY P9K_NORDVPN_COUNTRY_CODE
+prompt_nordvpn:2> [[ -e /run/nordvpn/nordvpnd.sock ]]
+prompt_nordvpn:3> _p9k_fetch_nordvpn_status # this hangs forever
```
https://github.com/romkatv/powerlevel10k/issues/2860

ruffsl reviewed (approved):
> add nordvpn.desktop to fix nordvpn login

I can confirm that login with desktop link integration works as expected via browser on KDE!

> While I try to get meshnet up and running, I'd like to avoid blocking on it and add the change in the next PR.

I've fine with meshnet in the next PR and getting the current features here merged.

My test branch for reference:
- https://github.com/SanferDsouza/nixpkgs/compare/nordvpn...ruffsl:nordvpn-ruffsl
- b7625f3f8be213533759d90e0c17abec10fba3fa

dimkNevidimk commented:
Unrelated to meshnet:

Does google search works ok for you in firefox (possibly in other browsers), when you connect to nordvpn?

For me searching in google in Firefox timeouts when I'm connected to any `nordvpn` server. Opening some sites work, like nordvpn.com, however e.g. github.com / google.com seem to timeout.
Note that  curl works ok for me, this is very weird.

I tried switching VPN technology to `OpenVPN` from `NordLynx` and it seems to fix the issue:
```bash
nordvpn set technology openvpn
```

Have you experienced anything similar? I tried with the latest commit from https://github.com/SanferDsouza/nixpkgs/tree/nordvpn.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5588

SanferDsouza commented:
> > hmm I'm experiencing the same issue too. `curl ipinfo.io` works but `curl www.google.com` doesn't work. Looking into it
> 
> Are you sure it wasn't from just picking a server was either a poor IP address reputation or being geo/bot blocked?
> 
> Google and GitHub services seem to be working fine from Chicago, regardless of browser or CLI tool.
> 
> ```
> > nordvpn status
> Status: Connected
> Server: United States #8798
> Hostname: us8798.nordvpn.com
> IP: ...
> Country: United States
> City: Chicago
> Current technology: NORDLYNX
> Current protocol: UDP
> Post-quantum VPN: Disabled
> Transfer: 33.64 MiB received, 1.86 MiB sent
> Uptime: 6 minutes 18 seconds
> ```
> 
> Could've also been a resuduall network complication from Google's global scale outage from last week?

Turns out it works for me! I had to disconnect from the vpn on my HOST system (and disable killswitch). After that the connection on my VM worked as expected. This is new behavior for me because all my previous testing was performed in the VM while also connected to vpn on the HOST. Anyway, `curl www.google.com` is working and I'm able to navigate the internet while connected to vpn using either protocols. @dimkNevidimk could this resolves your issue?

LeSuisse commented on pkgs/by-name/no/nordvpn/package.nix:
What's the purpose of this salt? It seems strange to have a static value shared between all users for something that is called a "salt".

SanferDsouza replied:
wrt their configuration file encryption practices, I opened an [issue upstream](https://github.com/NordSecurity/nordvpn-linux/issues/967) and currently await their reply.

SanferDsouza replied:
Upstream responded with the following:

> The use of SALT will be removed, most likely in the next major release. This decision is based on the fact that the encrypted files are already safeguarded by operating system permissions—only the root user or members of the nordvpn group can access them. Additionally, since the SALT is included in the build variables, it can be easily accessed, reducing its effectiveness as a security measure.
This is a known thing and because of this there are already comments in the code indicating that SALT is deprecated.

> The change will most likely be part of a major release because it breaks backward compatibility. Once the application is updated without SALT and configuration files changed, the older versions will not be able to read the files anymore.

I think we can rely on the filesystem permissions and consequently ignore the security implications of the deprecated salt.

LeSuisse commented on pkgs/by-name/no/nordvpn/package.nix:
I suggest to only mention `"loose"` to not encourage users to completely drop it.

Blatzar commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,200 @@
+    in
+    appendOverlays [
+      (final: prev: {
+        # Nordvpn uses a patched openvpn in order to perform xor obfuscation
```
As far as I understand they have not been integrated into OpenVPN as they are simply subpar/insecure when better alternatives exist. 
See https://github.com/OpenVPN/openvpn/pull/7#issuecomment-220348425 and https://tunnelblick.net/cOpenvpn_xorpatch.html

The patches basically allow you to scramble the traffic in order to bypass VPN blocks using basic packet inspection.

SanferDsouza replied:
I fixed the concern of "outdated urls" by storing the patches locally.

LeSuisse replied:
Since we can pull the patches from a repository that is the known source for it and that is unlikely to disappear I would suggest pulling the patches from it. It saves a few KB in the nixpkgs repository and it makes the origin clear. 

Also, given we have a patch series to maintain you can probably pull the whole tunnelblick repository from their stable tag like [v8.0](https://github.com/Tunnelblick/Tunnelblick/releases/tag/v8.0) and retrieve the patches from there. This way we do not have to get each patch individually and it is reasonably easy to get the latest version from a point in time where they have tested it. It is also make things easier to write an automated update script.

SanferDsouza commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,200 @@
+let
+  patchedPkgs =
+    let
+      baseURL = "https://github.com/Tunnelblick/Tunnelblick/raw/master/third_party/sources/openvpn";
```
I fixed this concern by storing the patches locally.

drupol commented:
> set version to 4.0.0

Don't forget to update the commit message accordingly.

ruffsl commented on pkgs/by-name/no/nordvpn/package.nix:
```diff
@@ -0,0 +1,170 @@
+  ];
+
+  # libxml2 2.14.[0-4] breaks daemon
+  libxml2' = libxml2.overrideAttrs (oldAttrs: {
```
Related forthcoming changes: 
- https://github.com/NixOS/nixpkgs/pull/421740

alyssais commented on pkgs/by-name/no/nordvpn/package.nix:
@LordGrimmauld ?

LordGrimmauld replied:
The CVE is tagged fixed in https://gitlab.gnome.org/GNOME/libxml2/-/releases/v2.14.4#security only. The [commit](https://gitlab.gnome.org/GNOME/libxml2/-/commit/acbbeef9f5dcdcc901c5f3fa14d583ef8cfd22f0) was made May 27th, way after 2.13.8 was released. The code that was vulnerable (and was changed for 2.14.4) is already present in [2.13.8](https://gitlab.gnome.org/GNOME/libxml2/-/blob/v2.13.8/tree.c?ref_type=tags#L173). Not sure where you got that `2.12.6` number, it was only fixed in `2.14.4`.

SanferDsouza replied:
Ah woops, I made a mistake! I misunderstood the CVE as it said `unaffected from 0:2.12.5-7.el10_0`, and since `2.12.5-7.el10_0 < 2.12.6 < 2.13.8` in semver.. I now understand that does not follow. Thanks for clarifying!

Looks like [421740](https://github.com/NixOS/nixpkgs/pull/421740) resolves the CVE, I intend to use `libxml2_13` once the pr merges.

getreu commented:
When can we expect this PR in unstable?

LordGrimmauld commented:
whenever staging-next is merged to master and the channel advances

different-error commented:
I face some difficulty atm. I cannot build staging locally. Running the tests suite or building the VM either fails or consumes too much memory. You can find these changes in my fork [here](https://github.com/different-error/nixpkgs/tree/nordvpn-staging). Since I cannot build staging, I cannot push my changes to this PR. I intend to try every so often, but worst case I can change base back to master once libxml2_13 merges.

ruffsl commented:
> I intend to try every so often, but worst case I can change base back to master once libxml2_13 merges.

Looks like `libxml2_13` has made its way into `staging-next`. The PR to watch for when it finally merges into main:

- https://github.com/NixOS/nixpkgs/pull/432489

different-error commented:
Changes:
- add `machine.shutdown()` after each test
- use plasma6 instead of gnome
- use Tunnelblick v8.0 downloaded from github
- use `f1nd1ngn3m0` as the salt

Testing:
- run automated tests
- successful connection and `curl ipinfo.io` using both openvpn and nordlynx

philiptaron commented on maintainers/maintainer-list.nix:
```diff
@@ -22650,6 +22650,11 @@
   };
+  sanferdsouza = {
+    name = "Sanfer D'souza";
+    github = "sanferdsouza";
```
```suggestion
    github = "different-error";
```

philiptaron reviewed (commented):
I'd like to merge the package without the NixOS module, which I feel unable to review and merge. Would you mind splitting the PR so that we can get `nordvpn` into nixpkgs, while someone with more module experience can keep chipping on the somewhat complicated module definition?

Since this PR was opened, your GitHub ID has changed; mind updating that commit also?

different-error commented:
> I'd like to merge the package without the NixOS module, which I feel unable to review and merge. Would you mind splitting the PR so that we can get `nordvpn` into nixpkgs, while someone with more module experience can keep chipping on the somewhat complicated module definition?
> 
> Since this PR was opened, your GitHub ID has changed; mind updating that commit also?

Thank you for catching that. I intend to split the PR by Monday unless it gains the necessary approvals before then. If required, I could do it sooner - lmk, ty.

different-error commented:
Thanks to @giacomofm for finding the semantic bug involving the incorrect swapping of defaultUser and defaultGroup.

ruffsl reviewed (approved):
I've shifted to unstable for some unrelated KDE fixes, but cherry picked these latest commits. Appreciate the update using the added libxml2 package. Working well.

ruffsl commented on nixos/doc/manual/release-notes/rl-2511.section.md:
```diff
@@ -114,6 +115,8 @@
 
 - [SillyTavern](https://sillytavern.app/), LLM Frontend for Power Users. Available as [services.sillytavern](#opt-services.sillytavern.enable).
 
+- [NordVPN](https://github.com/NordSecurity/nordvpn-linux), a NordVPN client for linux. Available as [services.nordvpn](options.html#opt-services.nordvpn.enable).
```
I'm all for linting white spaces in text, but is it the norm to include them in package PRs?

different-error replied:
ah I didn't realize this had occurred during my rebase. Ty

ruffsl reviewed (approved):
As for the nix code, I've tested the latest changes to this package and module locally, and all's still building and running as expected.

ruffsl commented:
@different-error , looks like v4.2.0 has released. I tried it locally, and the CLI seems to be working well just the same: 

```diff
diff --git a/pkgs/nordvpn/package.nix b/pkgs/nordvpn/package.nix
index 0038a2d..b9c68c7 100644
--- a/pkgs/nordvpn/package.nix
+++ b/pkgs/nordvpn/package.nix
@@ -42,13 +42,13 @@ let
 in
 buildGoModule (finalAttrs: {
   pname = "nordvpn";
-  version = "4.0.0";
+  version = "4.2.0";
 
   src = fetchFromGitHub {
     owner = "NordSecurity";
     repo = "nordvpn-linux";
     tag = finalAttrs.version;
-    hash = "sha256-0GgMIFi6qrO7NG94vvWprwWr+8j87M5eH5W/pCtSWqs=";
+    hash = "sha256-9uh/UkOS84tVeW/d6qQ6bYPXzGXEoD21QHzrcMcdj7M=";
   };
 
   nativeBuildInputs = [
@@ -64,7 +64,7 @@ buildGoModule (finalAttrs: {
     libxml2_13
   ];
 
-  vendorHash = "sha256-GREoxjXqb28nabNvvcGjQA1rG4h9e/gINqEPY4++6Oo=";
+  vendorHash = "sha256-eUM69CQjbML8fWRG8H3w6x4M+E51YrXX/UCUFHerQmM=";
 
   modPostBuild = ''
     patch -p0 < ${./gokogiri-xpath-expression.patch}
```

Perhaps we could also start a separate PR to begin packaging the new GUI that NordVPN open sourced with v4.2.0:

- GUI Build
  - https://github.com/NordSecurity/nordvpn-linux/blob/4.2.0/gui/BUILD.md
  - https://github.com/NordSecurity/nordvpn-linux/blob/4.2.0/.github/workflows/gui_ci.yml
- v4.0.0 vs v4.2.0
  - https://github.com/NordSecurity/nordvpn-linux/compare/919fc28253c3ad802403ea156ea511ab2d131e40..84d6668db19105aa50be23f24cadd62da6eff995

different-error commented:
> @different-error , looks like v4.2.0 has released. I tried it locally, and the CLI seems to be working well just the same:

Thanks for the hashes.

> perhaps we could also start a separate PR to begin packaging the new GUI that NordVPN open sourced with v4.2.0:

Sounds good, I can (eventually..) start a draft PR or two for GUI and meshnet support. Looks like they use flutter and I can't find documentation for `buildFlutterApplication` but I can try to figure something out!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2601

ruffsl commented:
> Any insight about why meshnet is not supported?

@andersonjoseph , meshnet requires dependencies such as `libtelio`, related discussion from Jun 13th to 18th above:

- https://github.com/NixOS/nixpkgs/pull/406725#issuecomment-2971505476
- https://github.com/NixOS/nixpkgs/pull/406725#issuecomment-2985538661

different-error commented:
update:
- I've added libdrop and modified the nordvpn cli recipe to use the tags "telio" "drop".
- they've patched libdrop-go so I believe we can use Go1.24+

Surprisingly, I don't see "meshnet" in the nordvpn settings.

https://github.com/different-error/nixpkgs/tree/nordvpn-meshnet
https://github.com/different-error/nixpkgs/tree/nordvpn-meshnet/pkgs/by-name/li/libdrop
https://github.com/different-error/nixpkgs/tree/nordvpn-meshnet/pkgs/by-name/no/nordvpn

I've caught the flu. I intend to get back to this when I feel healthy enough to do so.

andersonjoseph commented:
> I could use some help getting meshnet working. Please base your changes around my feature branch `nordvpn-meshnet`. Ty

I will take a look when I get home from holiday travel (in a couple of hours) 👌

different-error committed:
nordvpn: init at 4.6.0

different-error committed:
nixos/nordvpn: init module

different-error committed:
nixos/nordvpn: init module

different-error commented:
Updated to version 4.5.0
